### PR TITLE
gCNV restarts inference with a new random seed in case ELBO converges to a NaN

### PR DESCRIPTION
**Pull request by asmirnov239**
This is intended to alleviate transient issues with GermlineCNVCaller inference in which the ELBO converges to a NaN value, by calling the python gCNV code with an updated random seed input.

### CODE REVIEW

**Review by cmnbroad — CHANGES_REQUESTED** [review]

**Comment by asmirnov239** [comment]
@cmnbroad I added the changes you suggested, could you take a look? 

I added a set of methods to `PythonScriptExecutor` mirroring existing ones that return booleans.  I tried to extract as much common code as possible. Also, because now the entry point in  `ScriptExecutor` that return `ProcessOutput` does not process and throw any exceptions, I extracted that processing code and manually call it from `GermlineCNVCaller`. Let me know if you have a better idea on a better way to handle it.

If that looks good to you, could you also suggest what unit tests you would like me to add?

**Review by cmnbroad — CHANGES_REQUESTED** [review]
@asmirnov239 Thanks for the update. There are a bunch of minor/cosmetic comments, and one request for a minor refactoring change, but overall this looks pretty good.

As for tests, I think there are 4 `PythonScriptExecutorUnit` test cases:

- Two cases for `executeScriptAndGetOutput(final Resource scriptResource,..), one that executes a script that just does `exit 239`, and one that is some simple code that just succeeds without setting an explicit exit code, with both calling and validating the `ProcessOutput` exitcode, and the return value from getExceptionMessageFromScriptError`.
- The same two cases, only for `executeArgsAndGetOutput`.

Ideally there would also be a negative case for GermlineCNVCaller if you can force it to fail to converge, but maybe thats harder.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java [hunk 1]
```diff
@@ -305,10 +310,30 @@ protected Object doWork() {
...
+        ProcessOutput pythonProcessOutput = executor.executeScriptAndGetOutput(
+                new Resource(script, GermlineCNVCaller.class),
+                null,
+                composePythonArguments(intervalSubsetReadCountFiles, randomGCNVSeed));
+        if (pythonProcessOutput.getExitValue() != 0) {
+            // We restart once if the inference diverged
+            if (pythonProcessOutput.getExitValue() == gcnvRestartExitCode) {
+                final Random generator = new Random(randomGCNVSeed);
+                final int nextGCNVSeed = generator.nextInt();
+                logger.info("The inference ran into a NaN error and will be restarted one more time.");
```
[the remark this answers is not in the document]
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:326]: Yes that's better I agree. Done

File: src/main/java/org/broadinstitute/hellbender/utils/python/PythonScriptExecutor.java [hunk 2]
```diff
@@ -105,12 +104,31 @@ public boolean executeModule(final String moduleName, final List<String> pythonP
...
+        }
+    }
+
     /**
      * Execute a python script from a Resource file.
      *
      * @param scriptResource {@link Resource} for the script to execute
      * @param pythonProcessArgs args to be passed to the python process
-     * @param scriptArgs args to be passed to the python code
+     * @param scriptArgs args to be passed to the python codescriptArgs
```
[the remark this answers is not in the document]
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/utils/python/PythonScriptExecutor.java:131]: Oops, this was an accident

File: src/main/resources/org/broadinstitute/hellbender/tools/copynumber/case_denoising_calling.py [hunk 3]
```diff
@@ -213,9 +214,14 @@ def update_args_dict_from_saved_model(input_model_path: str,
...
-    # go!
-    task.engage()
-    task.disengage()
+    try:
+        # go!
+        task.engage()
+        task.disengage()
+    except gcnvkernel.ConvergenceError as err:
+        logger.info(err.message)
+        sys.exit(239) # Pass an exit code to Java side indicating restart is needed
```
[the remark this answers is not in the document]
Comment (asmirnov239) [reply src/main/resources/org/broadinstitute/hellbender/tools/copynumber/case_denoising_calling.py:223]: Definitely. I added it to gcnvkernel.io_const and kept it lower case to stick with the convention there

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java [hunk 4]
```diff
@@ -289,6 +291,9 @@
     private SimpleIntervalCollection specifiedIntervals;
     private File specifiedIntervalsFile;
```
[the remark this answers is not in the document]
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:293]: Done!

File: src/main/java/org/broadinstitute/hellbender/utils/runtime/ScriptExecutor.java [hunk 5]
```diff
@@ -72,62 +74,52 @@ protected void executableMissing() {
...
      */
-    protected boolean executeCuratedArgs(final String[] commandLineArguments) {
-        if (!externalExecutableExists()) {
-            if (!ignoreExceptions) {
-                executableMissing();
-            } else {
-                logger.warn("Skipping: " + getApproximateCommandLine());
-                return false;
-            }
+    public String getExceptionMessageFromScriptError(final ProcessOutput po) {
```
[the remark this answers is not in the document]
Comment (asmirnov239) [reply src/main/java/org/broadinstitute/hellbender/utils/runtime/ScriptExecutor.java:82]: I like that. Done!

**Comment by asmirnov239** [comment]
@cmnbroad I addressed your comments, except for making a negative test case for GermlineCNVCaller - I added that to  #6893 issue with an ongoing list of needed tests. Let me know if the changes look good to you.

File: src/main/java/org/broadinstitute/hellbender/utils/python/PythonScriptExecutor.java [hunk 6]
```diff
@@ -205,4 +228,38 @@ public static void checkPythonEnvironmentForPackage(final String packageName) {
             throw new RuntimeException(errorMessage, e);
         }
     }
+
+    /**
+     * Auxiliary method that validates and builds python command line argument list
+     */
+    private List<String> validateAndBuildCommand(final String scriptName, final List<String> pythonProcessArgs, final List<String> scriptArgs) {
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/utils/python/PythonScriptExecutor.java:235]: This method can be static.

**Review by cmnbroad — APPROVED** [review]
@asmirnov239 This looks pretty good now - only one minor comment. I'm not sure if this has already had a merge test run with the changes in https://github.com/broadinstitute/gatk/pull/6554, so it might be worth rebasing and running the tests one more time before merge.

**Pull request merged**